### PR TITLE
Fix FA2 long-context perf regression (#46693)

### LINKED ISSUE #46693: Performance regression with FA2 and long-context generation after 5.2.0
### System Info

transformers: 5.2.0 vs 5.4.0 (I could not test 5.3.0 due to a bug)
PyTorch: 2.12
flash-attn: 2.8.3
CUDA: 13.2
Python: 3.12
cudnn: 9.20.0.48

### Who can help?

@zucchini-nlp 
@Cyrilvallez 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

1. Install the latest version of transformers
2. Install FlashAttention2
3. Run the provided code
4. Downgrade to transformers version 5.2.0
5. Run the code again

```python
from transformers import Qwen3VLForConditionalGeneration, AutoProcessor
import torch

model = Qwen3VLForConditionalGeneration.from_pretrained(
    "Qwen/Qwen3-VL-2B-Thinking",
    dtype=torch.bfloat16,
    attn_implementation="flash_attention_2",
).to("cuda")

processor = AutoProcessor.from_pretrained("Qwen/Qwen3-VL-2B-Thinking")

messages = [
    [
        {
            "role": "user",
            "content": [
                {"type": "video", "video": torch.zeros(768, 3, 512, 512)},
                {"type": "text", "text": "placeholder"},
            ],
        }
    ],
    [
        {
            "role": "user",
            "content": [
                {"type": "video", "video": torch.zeros(768, 3, 512, 512)},
                {"type": "text", "text": "different placeholder"},
            ],
        }
    ],
]

inputs = processor.apply_chat_template(
    messages,
    tokenize=True,
    add_generation_prompt=True,
    do_sample_frames=False,
    return_dict=True,
    return_tensors="pt",
    padding=True,
    padding_side="left",
).to("cuda")

MAX_LEN = 16 * 1024

while True:
    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)

    start.record()
    out_ids = model.generate(
        **inputs, do_sample=False, min_length=MAX_LEN, max_length=MAX_LEN, use_cache=True
    )
    end.record()

    torch.cuda.synchronize()
    assert out_ids.size(0) == 2 and out_ids.size(1) == MAX_LEN

    cur_time = start.elapsed_time(end)
    print(cur_time)
```

### Expected behavior

Tested on an NVIDIA RTX 6000 PRO:
- v5.2.0 FA2 = 92 sec
- v5.4.0 FA2 = 108 sec (+17% overhead)
- v5.2.0 SDPA = 136 sec
- v5.4.0 SDPA = 136 sec (no change)

The performance regression seems to occur when the KV cache is sufficiently large, at which point the GPU memory bandwidth has been saturated. Furthermore, it seems to be related with the changes to the FA2 integration layer. It does not seem to be related with the nvidia-cudnn conv kernel performance regression which was fixed in a recent PyTorch version.

### PR DESCRIPTION
**Pull request by ammmanism**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47215)

## 🔧 Fix ready — root cause confirmed + patched

Traced the regression down to `.item()` removal in **PR #42435**. Three call sites were left returning 0-dim CUDA tensors instead of Python ints for `max_seqlen_in_batch`:

- `_get_unpad_data()`
- `_unpad_input()`
- `prepare_fa_kwargs_from_position_ids()`

These tensors get passed straight into `flash_attn_varlen_func(max_seqlen_q=..., max_seqlen_k=...)`, and on bandwidth-saturated GPUs (🔴 RTX 6000 Pro, 🟡 milder on RTX 4090) with large KV caches, this triggers extra kernel-dispatch/sync overhead per decode step — compounding across thousands of steps into the ~17% slowdown reported here.

**Fix:** restored `.item()` at all three sites, gated behind `is_torchdynamo_compiling()` so `torch.compile` graph capture isn't broken — tensor form is preserved under compile, scalar form used in eager. The downstream `_process_flash_attention_kwargs()` now also converts tensors to ints in eager mode (not just during tracing), ensuring the scalar reaches the flash kernel.

### Verification
- ✅ Regression test asserting scalar type in eager mode: `tests/utils/test_flash_attention_utils.py` (3 tests pass)
- ✅ Standalone microbenchmark isolating the mechanism: `bench_fa2_unpad.py` (runs on CUDA; CPU shows no delta as expected, regression is GPU/bandwidth-specific)
- ✅ Compile-mode guard verified: under `torch.compile` the tensor form is preserved (no graph break); in eager mode a Python `int` is returned
- ✅ `ruff check` / `ruff format --check` clean; `utils/check_copies.py` passes

### Files changed
- `src/transformers/modeling_flash_attention_utils.py` — restore `.item()` with `is_torchdynamo_compiling()` guard in the three call sites + eager-mode conversion in `_process_flash_attention_kwargs`
- `tests/utils/test_flash_attention_utils.py` — regression test for scalar return type
- `bench_fa2_unpad.py` — microbenchmark

Fixes #46693

cc @vasqu @zucchini-nlp @andreasgoulas — building on your investigation and the draft in #47134, this should close the residual gap you flagged. Happy to adjust based on review.

### CODE REVIEW

**Comment by vasqu** [comment]
Hmm afaik, #47134 should suffice. Also there seem to be multiple different fixes being merged into this PR - maybe you built on top of other branches that included different fixes?

**Comment by ammmanism** [comment]
@vasqu Good catch, thanks. This branch picked up unrelated commits (tokenizer thread-safety, CI workflow permissions, an unrelated ruff import fix) — looks like it was cut from a stale/dirty local branch rather than clean main. That also explains the security gate trip and the code quality failure.
Re: #47134 — happy to close this in favor of that one if it already covers the fix; otherwise I'll rebuild this branch from clean main with just the FA2 commit + test + benchmark so the diff is scoped correctly (should be ~3 files, not 5+769 lines). Will push the cleaned-up version shortly.
